packages/sui-react-router/src/PatternUtils.js

```diff
@@ -0,0 +1,237 @@
+}
+
+/**
+ * Attempts to match a pattern on the given pathname. Patterns may use
+ * the following special characters:
+ *
+ * - :paramName     Matches a URL segment up to the next /, ?, or #. The
+ *                  captured string is considered a "param"
+ * - ()             Wraps a segment of the URL that is optional
+ * - *              Consumes (non-greedy) all characters up to the next
+ *                  character in the pattern, or to the end of the URL if
+ *                  there is none
+ * - **             Consumes (greedy) all characters up to the next character
+ *                  in the pattern, or to the end of the URL if there is none
+ *
+ * The function calls callback(error, matched) when finished.
+ * The return value is an object with the following properties:
+ *
+ * - remainingPathname
+ * - paramNames
+ * - paramValues
+ */
+export function matchPattern(pattern, pathname) {
+  // Ensure pattern starts with leading slash for consistency with pathname.
+  if (pattern.charAt(0) !== '/') {
+    pattern = `/${pattern}`
+  }
+  let {regexpSource, paramNames, tokens} = compilePattern(pattern)
+
+  if (pattern.charAt(pattern.length - 1) !== '/') {
+    regexpSource += '/?' // Allow optional path separator at end.
+  }
+
+  // Special-case patterns like '*' for catch-all routes.
+  if (tokens[tokens.length - 1] === '*') {
+    regexpSource += '$'
+  }
+
+  const match = pathname.match(new RegExp(`^${regexpSource}`, 'i'))
+  if (match == null) {
+    return null
+  }
+
+  const matchedPath = match[0]
+  let remainingPathname = pathname.substr(matchedPath.length)
+
+  if (remainingPathname) {
+    // Require that the match ends at a path separator, if we didn't match
+    // the full path, so any remaining pathname is a new path segment.
+    if (matchedPath.charAt(matchedPath.length - 1) !== '/') {
+      return null
+    }
+
+    // If there is a remaining pathname, treat the path separator as part of
+    // the remaining pathname for properly continuing the match.
+    remainingPathname = `/${remainingPathname}`
+  }
+
+  return {
+    remainingPathname,
+    paramNames,
+    paramValues: match.slice(1).map(v => v && decodeURIComponent(v))
+  }
+}
+
+export function getParamNames(pattern) {
+  return compilePattern(pattern).paramNames
+}
+
+export function getParams(pattern, pathname) {
+  const match = matchPattern(pattern, pathname)
+  if (!match) {
+    return null
+  }
+
+  const {paramNames, paramValues} = match
+  const params = {}
+
+  paramNames.forEach((paramName, index) => {
+    params[paramName] = paramValues[index]
+  })
+
+  return params
+}
+
+/**
+ * Returns a version of the given pattern with params interpolated. Throws
+ * if there is a dynamic segment of the pattern for which there is no param.
+ */
+export function formatPattern(pattern, params = {}) {
+  const {tokens} = compilePattern(pattern)
+  let parenCount = 0
+  let pathname = ''
+  let splatIndex = 0
+  const parenHistory = []
```

Comment:
👀 shouldn't it read `parentHistory`?

Reply:
This whole file is a copy & paste 🍝 of react-router v3 as noted in the first line and thus is getting all the names for the variables.

I noticed as well but keep it in order to mantain the same code in case we need to recheck something from the original codebase.

Reply:
oh got it! 👍